src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java

```diff
@@ -26,13 +32,50 @@
 public class RestUtil {
 
     private static final int MAX_IN_MEMORY_SIZE = 1 * 1024 * 1024; // 1MB
+    private final ObjectMapper objectMapper = new ObjectMapper();
 
-    public OAuthClient createOAuthClient(URL authorizationUrl) {
-        return new OAuthClient(authorizationUrl);
+    public OAuthClient createOAuthClient(URL controllerUrl, ConnectionContext connectionContext, String origin,
+                                         boolean shouldTrustSelfSignedCertificates) {
+        Map<String, Object> infoMap = getInfoMap(controllerUrl, createWebClient(shouldTrustSelfSignedCertificates));
+        URL authorizationUrl = getAuthorizationUrl(infoMap);
+        return new OAuthClientWithLoginHint(authorizationUrl, connectionContext, origin, createWebClient(true));
     }
 
-    public OAuthClient createOAuthClient(URL authorizationUrl, ConnectionContext connectionContext, String origin) {
-        return new OAuthClientWithLoginHint(authorizationUrl, connectionContext, origin);
+    public OAuthClient createOAuthClientByControllerUrl(URL controllerUrl, boolean shouldTrustSelfSignedCertificates) {
+        WebClient webClient = createWebClient(shouldTrustSelfSignedCertificates);
+        URL authorizationUrl = getAuthorizationUrl(controllerUrl, webClient);
+        return new OAuthClient(authorizationUrl, webClient);
+    }
+
+    private URL getAuthorizationUrl(URL controllerUrl, WebClient webClient) {
+        Map<String, Object> infoMap = getInfoMap(controllerUrl, webClient);
+        return getAuthorizationUrl(infoMap);
+    }
+
+    // TODO: Refactor not to use v2
+    private Map<String, Object> getInfoMap(URL controllerUrl, WebClient webClient) {
+        String infoResponse = webClient.get()
+                                       .uri(controllerUrl + "/v2/info")
+                                       .retrieve()
+                                       .bodyToMono(String.class)
+                                       .block();
+        try {
+            return objectMapper.readValue(infoResponse, new TypeReference<Map<String, Object>>() {
+            });
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Error getting /v2/info from cloud controller.", e);
+        }
+    }
+
+    private URL getAuthorizationUrl(Map<String, Object> infoMap) {
+        String authorizationEndpoint = (String) infoMap.get("authorization_endpoint");
```

Comment:
The v3 information about the authorization url is changed.
See this commit https://github.com/cloudfoundry-incubator/multiapps-controller/commit/e96f7526ff9103a9c28dc868c31f9e9895a26b7e for more information

Reply:
Reused the code